### ISSUE TITLE
Replace printf

### ISSUE THREAD

**Issue by cvonder1** [lmaraite/tcp-server#33]
In the code are multiple `printf` statements in asynchronous parts. The documentation of `fork` states, that functions such as `printf` are not thread-safe([here](https://www.man7.org/linux/man-pages/man2/fork.2.html) and [here](https://www.man7.org/linux/man-pages/man7/signal-safety.7.html)).
If such logging mechanism is still needed, a more secure way should be elaborated. Maybe it's reasonable to include a logging library or write our own logging methods.

Part of #32

**Comment by cvonder1** [comment]
duplicate of #37

**Comment by lmaraite** [comment]
try out sprintf + write(stdout)

**Comment by cvonder1** [comment]
This really isn't that clear, if `printf` is thread-safe or not.
There are two contradicting documentations (at least in my head):

First there is `man 3 printf`, which states that `printf` is only MT-Safe locale. I think, that MT-Safe locale means, that `printf` can produce gibberish, when called concurrently, but will not crash the program. That was elaborated a little [here](https://github.com/lmaraite/tcp-server/pull/41#discussion_r650848445).

On the other hand there is `man 3 flockfile`. `flockfile` is kind of a semaphore for the stdout and stdin. The documentation also states, that the standard input/output functions(such as `printf`) have to use `flockfile`. Therefore they would be completely thread-safe.

I did some experiments with the code snippet at the end of this post. Running those tests today, showed that every line is intact, when the standard buffering mechanism is used(so the buffer is flushed after each new line). No buffer at all activated with `no_buffer` or with the `write` function work as well. Just using the full buffer(activated with full_buffer) didn't work. I am not really certain why this is so.

The implementation of `printf` can be found [here](https://code.woboq.org/userspace/glibc/stdio-common/printf.c.html) and the internal implementation [here](https://code.woboq.org/userspace/glibc/stdio-common/vfprintf-internal.c.html#1245). This glibc code is really messy, so I hope I really found the correct parts.
In [line 1327](https://code.woboq.org/userspace/glibc/stdio-common/vfprintf-internal.c.html#1327) there is the _IO_lockfile. That is a hell of a macro and I couldn't find the concrete implementation. So I just assume it is more or less the `flockfile`. The output is then printed with the `output`, which uses another macro `PUT`, which uses an other macro and so on... (They even use a macro called JUMP, which will jump to a given pointer...).
tldr: It's still unclear to me how the `printf` function works. The experiments show, that the default buffering works with multiple threads. So we should be good for now. Eventually one could implement a logging utility secured by `flockfile` and with additional logging level support. There is further tests needed, if `flockfile` does the job for completely buffered stdout.

I think, this is not a major problem, so if there is something more urgent, I can do that instead.

```
#include <unistd.h>
#include <sys/types.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>

void no_buffer() {
    setvbuf(stdout, NULL, _IONBF, 0);
}

void full_buffer() {
    char* buffer = malloc(sizeof(char) * 1024);
    memset(buffer, '\0', sizeof(char) * 1024);
    setvbuf(stdout, buffer, _IOFBF, 1024);
}

int main() {
    //no_buffer();
    full_buffer();

    char i;
    for(i = 'A'; i < 'z'; i++) {
        if(fork() == 0) {
            break;
        }
    }
    printf("child: %c\n", i);

    char output[100];
    for(int j = 0; j < 100; j++) {
        output[j] = i;
    }
    output[99]='\n';
    output[100] = '\0';

    for(int j = 0; j < 1000; j++) {
        //write(1, output, 100);
        printf("%s", output);
    }

    return 0;
}
```
